tests.py

```diff
@@ -1,107 +1,192 @@
+#!/usr/bin/python
```

Comment:
What if we kept a dataset, generated with the original code, and had a `nose` unit test that runs the current implementation and compares the results to the known results.